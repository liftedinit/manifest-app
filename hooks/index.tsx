export * from './useFeeEstimation';
export * from './useIpfs';
export * from './useLcdQueryClient';
export * from './usePoaLcdQueryClient';
export * from './useQueries';
export * from './useTx';
export * from './useContacts';
export { default as useIsMobile } from './useIsMobile';
