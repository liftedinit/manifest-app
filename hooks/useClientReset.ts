import { useQueryClient } from '@tanstack/react-query';
import { useCallback, useContext } from 'react';

import { Web3AuthContext } from '@/contexts/web3AuthContext';

/**
 * Hook to force complete reset of cosmos-kit and Web3Auth clients.
 * This completely re-mounts the ChainProvider to clear all cached state.
 */
export const useClientReset = () => {
  const { resetWeb3AuthClients, forceChainProviderReset } = useContext(Web3AuthContext);
  const queryClient = useQueryClient();

  const forceCompleteReset = useCallback(async () => {
    if (forceChainProviderReset) {
      // Reset Web3Auth state
      await resetWeb3AuthClients();

      // Clear all caches
      queryClient.clear();

      // Force complete re-mount
      forceChainProviderReset();

      return true;
    } else {
      return false;
    }
  }, [forceChainProviderReset, resetWeb3AuthClients, queryClient]);

  return {
    forceCompleteReset,
  };
};
