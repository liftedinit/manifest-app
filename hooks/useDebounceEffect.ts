import { DependencyList, useEffect } from 'react';

/**
 * Create a debounce effect that will call the callback function after the delay has passed,
 * on an interval.
 * @param callback The function to call
 * @param delay The delay in milliseconds
 * @param dependencies The dependencies to watch for changes
 */
export const useIntervalDebounceEffect = (
  callback: () => Promise<unknown>,
  delay: number,
  dependencies?: DependencyList
) => {
  useEffect(() => {
    let done = false;
    let latestTimer: Timer | undefined;

    async function inner() {
      if (done) return;

      try {
        await callback();
      } catch (error) {
        console.error('Error during refetch:', error);
      } finally {
        if (!done) {
          latestTimer = setTimeout(inner, delay);
        }
      }
    }

    latestTimer = setTimeout(inner, delay);

    return () => {
      done = true;
      clearTimeout(latestTimer);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, dependencies);
};
