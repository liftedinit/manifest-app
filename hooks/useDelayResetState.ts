import { Dispatch, SetStateAction, useCallback, useEffect, useRef, useState } from 'react';

/**
 * Create a state that will reset after a delayed to its initial or previous value.
 * Can be updated to a new value without resetting.
 * @param initialState
 * @param delay
 * @returns [state, setState, updateState]
 */
export function useDelayResetState<T>(
  initialState: T,
  delay: number
): [T, Dispatch<SetStateAction<T>>, Dispatch<SetStateAction<T>>] {
  const [state, setState] = useState(initialState);
  const resetStateRef = useRef<T>(initialState);
  const timerRef = useRef<Timer | undefined>();

  const setDelayResetState = (value: SetStateAction<T>) => {
    if (typeof value === 'function') {
      value = (value as (prevState: T) => T)(state);
    }
    if (timerRef.current) {
      clearTimeout(timerRef.current);
    }

    setState(value);
    timerRef.current = setTimeout(() => {
      setState(resetStateRef.current);
      clearTimeout(timerRef.current);
      timerRef.current = undefined;
    }, delay);
  };

  const updateBaseState = (value: SetStateAction<T>) => {
    if (typeof value === 'function') {
      value = (value as (prevState: T) => T)(state);
    }
    resetStateRef.current = value;
  };

  return [state, setDelayResetState, updateBaseState] as const;
}
