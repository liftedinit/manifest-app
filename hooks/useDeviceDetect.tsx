import { useState, useEffect } from 'react';
declare global {
  interface Navigator {
    userAgentData?: {
      mobile: boolean;
    };
  }
}
export const useDeviceDetect = () => {
  const [isMobile, setIsMobile] = useState(false);

  useEffect(() => {
    const checkDevice = () => {
      if (navigator.userAgentData) {
        setIsMobile(navigator.userAgentData.mobile);
      } else {
        setIsMobile(
          /Android|webOS|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini/i.test(navigator.userAgent)
        );
      }
    };

    checkDevice();
  }, []);

  return { isMobile };
};
