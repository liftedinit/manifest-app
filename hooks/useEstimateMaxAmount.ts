import BigNumber from 'bignumber.js';

import { CombinedBalanceInfo, isMfxToken } from '@/utils';

/**
 * The constant fee for the UMFX token.
 */
export const MFX_FEES_CONSTANT = 0.1;

/**
 * Hook to estimate the maximum token amount that can be sent.
 */
export function useEstimateMaxTokenAmount() {
  return (selectedToken: CombinedBalanceInfo): BigNumber => {
    const exponent = selectedToken.metadata?.denom_units[1]?.exponent ?? 6;
    const maxAmount = new BigNumber(selectedToken.amount).div(Math.pow(10, exponent));

    let adjustedMaxAmount = maxAmount;
    if (isMfxToken(selectedToken.base)) {
      if (maxAmount.gt(MFX_FEES_CONSTANT)) {
        adjustedMaxAmount = maxAmount.minus(MFX_FEES_CONSTANT);
      } else {
        adjustedMaxAmount = new BigNumber(0);
      }
    }
    return adjustedMaxAmount;
  };
}
