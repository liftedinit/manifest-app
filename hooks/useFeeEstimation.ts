import { EncodeObject } from '@cosmjs/proto-signing';
import { GasPrice, calculateFee } from '@cosmjs/stargate';
import { useChain } from '@cosmos-kit/react';
import { useToast } from '@/contexts/toastContext';

export const useFeeEstimation = (chainName: string) => {
  const { getSigningStargateClient, chain } = useChain(chainName);
  const { setToastMessage } = useToast();

  const gasPrice = chain.fees?.fee_tokens[0].average_gas_price || 0.025;

  const estimateFee = async (
    address: string,
    messages: EncodeObject[],
    modifier?: number,
    memo?: string
  ) => {
    try {
      const stargateClient = await getSigningStargateClient();
      if (!stargateClient) {
        setToastMessage({
          type: 'alert-error',
          title: 'Fee Estimation Error',
          description: 'Unable to connect to the network. Please try again.',
          bgColor: '#e74c3c',
        });
        return null;
      }

      // Check if account exists before simulation
      try {
        const account = await stargateClient.getAccount(address);
        if (!account) {
          setToastMessage({
            type: 'alert-error',
            title: 'Account Error',
            description: `Account ${address} does not exist on chain. Please ensure you have funds in your wallet.`,
            bgColor: '#e74c3c',
          });
          return null;
        }
      } catch (accountError) {
        setToastMessage({
          type: 'alert-error',
          title: 'Account Error',
          description: `Account ${address} does not exist on chain. Please ensure you have funds in your wallet.`,
          bgColor: '#e74c3c',
        });
        return null;
      }

      const gasEstimation = await stargateClient.simulate(address, messages, memo);
      if (!gasEstimation) {
        setToastMessage({
          type: 'alert-error',
          title: 'Fee Estimation Error',
          description: 'Failed to estimate transaction fees. Please try again.',
          bgColor: '#e74c3c',
        });
        return null;
      }

      // Ensure we have a valid gas price
      if (!gasPrice) {
        setToastMessage({
          type: 'alert-error',
          title: 'Configuration Error',
          description: 'Gas price configuration is missing. Please try again later.',
          bgColor: '#e74c3c',
        });
        return null;
      }

      return calculateFee(
        Math.round(gasEstimation * (modifier || 1.5)),
        GasPrice.fromString(`${gasPrice}umfx`)
      );
    } catch (error: any) {
      setToastMessage({
        type: 'alert-error',
        title: 'Fee Estimation Error',
        description: error.message || 'Failed to estimate transaction fees. Please try again.',
        bgColor: '#e74c3c',
      });
      return null;
    }
  };

  return { estimateFee };
};
