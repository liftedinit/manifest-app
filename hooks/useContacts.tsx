import { useCallback, createContext, ReactNode, useContext } from 'react';
import { Contact } from '@/utils/types';
import Yup from '@/utils/yupExtensions';
import { useLocalStorage } from '@/hooks/useLocalStorage';

const STORAGE_KEY = 'manifest.contacts';

const ContactSchema = Yup.object().shape({
  name: Yup.string().required('Name is required').noProfanity(),
  address: Yup.string().required('Address is required').manifestAddress(),
});

const ContactsArraySchema = Yup.array().of(ContactSchema);

export type ContactIndex = Partial<Contact> | number;

export interface ContactsContextType {
  contacts: Contact[];
  addContact: (contact: Contact) => boolean;
  updateContact: (index: ContactIndex, contact: Contact) => void;
  removeContact: (index: ContactIndex) => void;
  importContacts: (contacts: Contact[]) => Promise<boolean>;
  exportContacts: () => string;
}

export const ContactsContext = createContext<ContactsContextType>({
  contacts: [],
  addContact: () => false,
  updateContact: () => {},
  removeContact: () => {},
  importContacts: async () => false,
  exportContacts: () => '{}',
});

export const ContactsProvider = ({ children }: { children: ReactNode }) => {
  const [contacts, setContacts] = useLocalStorage<Contact[]>(STORAGE_KEY, []);

  const findIndex = (index: ContactIndex) => {
    return typeof index === 'number'
      ? index
      : contacts.findIndex(c => c.name === index.name || c.address === index.address);
  };

  const updateContact = (index: ContactIndex, updatedContact: Contact) => {
    const toUpdate = findIndex(index);
    const c = ContactSchema.validateSync(updatedContact);

    setContacts(contacts.map((contact, i) => (i === toUpdate ? c : contact)));
  };

  const addContact = (contact: Contact, rethrow = false) => {
    try {
      contact = ContactSchema.validateSync(contact);
    } catch (error) {
      if (rethrow) {
        throw error;
      }
      return false;
    }

    if (contacts.some(c => c.address === contact.address)) {
      updateContact(contact, contact);
    } else {
      setContacts([...contacts, contact]);
    }
    return true;
  };

  const removeContact = (index: ContactIndex) => {
    const toRemove = findIndex(index);
    const newContacts = contacts.filter((_, i) => i !== toRemove);
    setContacts(newContacts);
  };

  const importContacts = async (newContacts: Contact[]) => {
    try {
      await ContactsArraySchema.validate(newContacts);

      for (const contact of newContacts) {
        addContact(contact);
      }
      return true;
    } catch (error) {
      console.error('Invalid contacts format', error);
      return false;
    }
  };

  const exportContacts = () => JSON.stringify(contacts);

  return (
    <ContactsContext.Provider
      value={{ contacts, addContact, removeContact, updateContact, importContacts, exportContacts }}
    >
      {children}
    </ContactsContext.Provider>
  );
};

export function useContacts() {
  return useContext(ContactsContext);
}
