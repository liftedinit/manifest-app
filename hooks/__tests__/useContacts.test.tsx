import { cleanup, fireEvent, render, waitFor } from '@testing-library/react';
import { afterEach, describe, expect, spyOn, test } from 'bun:test';
import React from 'react';

import { ContactsProvider, STORAGE_KEY, useContacts } from '@/hooks';
import { formatComponent } from '@/tests';

function TestComponent() {
  const { contacts, addContact, removeContact } = useContacts();

  return (
    <>
      <div>
        <div data-testid="contact-count">{contacts.length}</div>
        <div data-testid="contact-list">
          {contacts.map((contact, index) => (
            <div key={index}>
              <div data-testid={`contact-name-${index}`}>{contact.name}</div>
              <div data-testid={`contact-address-${index}`}>{contact.address}</div>
              <div>
                <button
                  data-testid={`contact-remove1-${index}`}
                  onClick={() => removeContact(index)}
                >
                  Remove by index
                </button>
                <button
                  data-testid={`contact-remove2-${index}`}
                  onClick={() => removeContact({ name: contact.name })}
                >
                  Remove by name
                </button>
              </div>
            </div>
          ))}
        </div>
      </div>

      <div>
        <input data-testid="contact-name" />
        <input data-testid="contact-address" />
        <button
          data-testid="contact-add"
          onClick={() => {
            addContact({
              name: (document.querySelector('[data-testid="contact-name"]') as HTMLInputElement)
                .value,
              address: (
                document.querySelector('[data-testid="contact-address"]') as HTMLInputElement
              ).value,
            });
          }}
        >
          Add
        </button>
      </div>
    </>
  );
}

describe('useContacts', () => {
  afterEach(() => {
    cleanup();
    window.localStorage.clear();
  });

  test('works', async () => {
    const wrapper = render(
      <ContactsProvider>
        <TestComponent />)
      </ContactsProvider>
    );
    expect(wrapper.getByTestId('contact-count')).toHaveTextContent('0');

    (wrapper.getByTestId('contact-name') as HTMLInputElement).value = 'Alice';
    (wrapper.getByTestId('contact-address') as HTMLInputElement).value =
      'manifest1aucdev30u9505dx9t6q5fkcm70sjg4rh7rn5nf';
    fireEvent.click(wrapper.getByTestId('contact-add'));

    wrapper.rerender(
      <ContactsProvider>
        <TestComponent />)
      </ContactsProvider>
    );

    await waitFor(() => {
      expect(wrapper.getByTestId('contact-count')).toHaveTextContent('1');
      expect(wrapper.getByTestId('contact-name-0')).toHaveTextContent('Alice');
      expect(wrapper.getByTestId('contact-address-0')).toHaveTextContent(
        'manifest1aucdev30u9505dx9t6q5fkcm70sjg4rh7rn5nf'
      );
    });

    (wrapper.getByTestId('contact-name') as HTMLInputElement).value = 'Bob';
    (wrapper.getByTestId('contact-address') as HTMLInputElement).value =
      'manifest1hj5fveer5cjtn4wd6wstzugjfdxzl0xp8ws9ct';
    wrapper.getByTestId('contact-add').click();

    wrapper.rerender(
      <ContactsProvider>
        <TestComponent />)
      </ContactsProvider>
    );

    await waitFor(() => expect(wrapper.getByTestId('contact-count')).toHaveTextContent('2'));

    expect(formatComponent(wrapper.asFragment())).toMatchSnapshot();
    expect(JSON.parse(window.localStorage.getItem(STORAGE_KEY)!)).toMatchSnapshot();
  });

  test('only allows valid Bech32 addresses', async () => {
    // Disable console.error
    spyOn(console, 'error');

    const wrapper = render(
      <ContactsProvider>
        <TestComponent />
      </ContactsProvider>
    );
    expect(wrapper.getByTestId('contact-count')).toHaveTextContent('0');

    (wrapper.getByTestId('contact-name') as HTMLInputElement).value = 'Alice';
    (wrapper.getByTestId('contact-address') as HTMLInputElement).value = 'invalid Bech32';

    wrapper.getByTestId('contact-add').click();

    wrapper.rerender(
      <ContactsProvider>
        <TestComponent />)
      </ContactsProvider>
    );

    await waitFor(() => {
      expect(wrapper.getByTestId('contact-count')).toHaveTextContent('0');
    });
  });

  test('removes contact by index', async () => {
    const wrapper = render(
      <ContactsProvider>
        <TestComponent />)
      </ContactsProvider>
    );
    expect(wrapper.getByTestId('contact-count')).toHaveTextContent('0');

    (wrapper.getByTestId('contact-name') as HTMLInputElement).value = 'Alice';
    (wrapper.getByTestId('contact-address') as HTMLInputElement).value =
      'manifest1aucdev30u9505dx9t6q5fkcm70sjg4rh7rn5nf';
    wrapper.getByTestId('contact-add').click();

    wrapper.rerender(
      <ContactsProvider>
        <TestComponent />)
      </ContactsProvider>
    );

    await waitFor(() => {
      expect(wrapper.getByTestId('contact-count')).toHaveTextContent('1');
    });

    wrapper.getByTestId('contact-remove1-0').click();

    wrapper.rerender(
      <ContactsProvider>
        <TestComponent />)
      </ContactsProvider>
    );

    await waitFor(() => {
      expect(wrapper.getByTestId('contact-count')).toHaveTextContent('0');
    });
  });

  test('adding existing address updates it', async () => {
    const wrapper = render(
      <ContactsProvider>
        <TestComponent />)
      </ContactsProvider>
    );
    expect(wrapper.getByTestId('contact-count')).toHaveTextContent('0');

    (wrapper.getByTestId('contact-name') as HTMLInputElement).value = 'Alice';
    (wrapper.getByTestId('contact-address') as HTMLInputElement).value =
      'manifest1aucdev30u9505dx9t6q5fkcm70sjg4rh7rn5nf';
    wrapper.getByTestId('contact-add').click();

    wrapper.rerender(
      <ContactsProvider>
        <TestComponent />)
      </ContactsProvider>
    );

    await waitFor(() => {
      expect(wrapper.getByTestId('contact-count')).toHaveTextContent('1');
      expect(wrapper.getByTestId('contact-name-0')).toHaveTextContent('Alice');
      expect(wrapper.getByTestId('contact-address-0')).toHaveTextContent(
        'manifest1aucdev30u9505dx9t6q5fkcm70sjg4rh7rn5nf'
      );
    });

    (wrapper.getByTestId('contact-name') as HTMLInputElement).value = 'Not Alice';
    (wrapper.getByTestId('contact-address') as HTMLInputElement).value =
      'manifest1aucdev30u9505dx9t6q5fkcm70sjg4rh7rn5nf';
    wrapper.getByTestId('contact-add').click();

    wrapper.rerender(
      <ContactsProvider>
        <TestComponent />)
      </ContactsProvider>
    );

    await waitFor(() => {
      expect(wrapper.getByTestId('contact-count')).toHaveTextContent('1');
      expect(wrapper.getByTestId('contact-name-0')).toHaveTextContent('Not Alice');
      expect(wrapper.getByTestId('contact-address-0')).toHaveTextContent(
        'manifest1aucdev30u9505dx9t6q5fkcm70sjg4rh7rn5nf'
      );
    });
  });
});
