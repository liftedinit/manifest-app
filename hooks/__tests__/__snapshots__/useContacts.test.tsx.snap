// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`useContacts works 1`] = `
"<DocumentFragment>
  <div>
    <div
      data-testid="contact-count"
    >
      2
    </div>
    <div
      data-testid="contact-list"
    >
      <div>
        <div
          data-testid="contact-name-0"
        >
          Alice
        </div>
        <div
          data-testid="contact-address-0"
        >
          manifest1aucdev30u9505dx9t6q5fkcm70sjg4rh7rn5nf
        </div>
        <div>
          <button
            data-testid="contact-remove1-0"
          >
            Remove by index
          </button>
          <button
            data-testid="contact-remove2-0"
          >
            Remove by name
          </button>
        </div>
      </div>
      <div>
        <div
          data-testid="contact-name-1"
        >
          Bob
        </div>
        <div
          data-testid="contact-address-1"
        >
          manifest1hj5fveer5cjtn4wd6wstzugjfdxzl0xp8ws9ct
        </div>
        <div>
          <button
            data-testid="contact-remove1-1"
          >
            Remove by index
          </button>
          <button
            data-testid="contact-remove2-1"
          >
            Remove by name
          </button>
        </div>
      </div>
    </div>
  </div>
  <div>
    <input
      data-testid="contact-name"
    />
    <input
      data-testid="contact-address"
    />
    <button
      data-testid="contact-add"
    >
      Add
    </button>
  </div>
  )
</DocumentFragment>"
`;

exports[`useContacts works 2`] = `
[
  {
    "address": "manifest1aucdev30u9505dx9t6q5fkcm70sjg4rh7rn5nf",
    "name": "Alice",
  },
  {
    "address": "manifest1hj5fveer5cjtn4wd6wstzugjfdxzl0xp8ws9ct",
    "name": "Bob",
  },
]
`;
